SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,493 @@
+			this.httpProtocol = protocol;
+			this.httpMethod = method;
+			this.sendLimit = sendLimit;
+			this.bufferSize = bufferSize;
+			this.byteLimitGet = byteLimitGet;
+			this.byteLimitPost = byteLimitPost;
+		}
+
+		/// <summary>
+		/// Adds an event payload to the database.
+		/// </summary>
+		/// <param name="payload">Payload.</param>
+		public void Add(TrackerPayload payload) {
+			this.eventStore.AddEvent (payload);
+			lock (emitLock) {
+				Monitor.Pulse(emitLock);
+			}
+		}
+
+		/// <summary>
+		/// Starts the Emitter; will be run every 10ms.
+		/// </summary>
+		public void Start() {
+			if (emitterTimer == null) {
+				Log.Debug("Emitter: Starting emitter.");
+				emitterTimer = new Timer(Emit, null, TIMER_INTERVAL, Timeout.Infinite );
+			}
+		}
+		
+		/// <summary>
+		/// Stops the Emitter; will also empty the queue into the database.
+		/// </summary>
+		public void Stop() {
+			if (emitterTimer != null) {
+				Log.Debug("Emitter: Stoppping emitter.");
+				emitterTimer.Change(Timeout.Infinite, Timeout.Infinite);
+				emitterTimer.Dispose();
+				emitterTimer = null;
+
+				lock (emitLock) {
+					Monitor.Pulse(emitLock);
+				}
+			}
+		}
+	
+		/// <summary>
+		/// Performs and orchestrates all sending of events from the database
+		/// and subsequent removal of events from the database.
+		/// </summary>
+		private void Emit(object state) {
+			sending = true;
+			List<EventRow> events = eventStore.GetDescEventRange (sendLimit);
+
+			// If some events were found start sending!
+			if (events.Count > 0) {
+				Log.Debug ("Emitter: Event count: " + events.Count);
+				List<RequestResult> results = SendRequests (events);
+				events = null;
+
+				int success = 0;
+				int failure = 0;
+
+				List<int> eventsToDelete = new List<int>();
+
+				foreach (RequestResult result in results) {
+					if (result.success) {
+						eventsToDelete.AddRange(result.rowIds);
+						success += result.rowIds.Count;
+					} else {
+						failure += result.rowIds.Count;
+					}
+				}
+
+				eventStore.DeleteEvents(eventsToDelete);
+
+				Log.Debug ("Emitter: event sending results.");
+				Log.Debug (" + Successful: " + success);
+				Log.Debug (" + Failure: " + failure);
+
+				if (failure > 0 && success == 0) {
+					// All events failed; sleep for 10 seconds and then retry.
+					// - Sleep prevents thrashing of database in situations where nothing can send.
+					Log.Error ("Emitter: All events failed to send; pausing emitter for ten seconds.");
+					sending = false;
+					if (emitterTimer != null) {
+						emitterTimer.Change (FAIL_INTERVAL, Timeout.Infinite);
+					}
+				} else {
+					if (emitterTimer != null) {
+						Log.Debug ("Emitter: All events sent successfully; checking for more...");
+						Emit (null);
+					}
+				}
+			} else {
+				Log.Debug ("Emitter: No more events to send; waiting for more.");
+				sending = false;
+				lock (emitLock) {
+					Monitor.Wait(emitLock);
+				}
+				if (emitterTimer != null) {
+					Emit (null);
```

Comment:
There's a potential stack overflow in this recursion. Do you think we should get rid of the Timer altogether and replace it with `while` and `Thread.sleep`?